### DISCUSSION TITLE
How to extract the image in JPXDecode

### DISCUSSION THREAD

**Discussion by thaihaquang** [pymupdf/PyMuPDF#1205] [Looking for help]
Dear authors,
I am trying your tool to extract images from PDF using the below code

<pre><code>
for i in range(len(doc)):
    for img in doc.getPageImageList(i):
        print("Processing ", i)
        xref = img[0]
        print (img)
        pix = fitz.Pixmap(doc, xref)
        if pix.n < 5:       # this is GRAY or RGB
            pix.writePNG("./out/p%s-%s.png" % (i, xref))
        else:               # CMYK: convert to RGB first
            pix1 = fitz.Pixmap(fitz.csRGB, pix)
            pix1.writePNG("./out/p%s-%s.png" % (i, xref))
            pix1 = None
        pix = None
</code></pre>

All files are extracted OK, except a PDF file.

Each image in that PDF are saved to 2 strange images
I found that the decode filter of the image is "JPXDecode"
Below is the information I got from doc.getPageImageList(i)
`[(85, 0, 1262, 1769, 8, 'DeviceRGB', '', 'Bg', 'JPXDecode'),
 (87, 86, 631, 885, 8, 'DeviceRGB', '', 'CL', 'JPXDecode')]`

Could you help me to explain the reason and guide me on how to fix this issue?
Thank you very much.

**Comment by JorjMcKie** [comment]
This is a scanned-in page.
Each scanner does this differently, so you may have a hard time to generate **_one_** identical looking image file from the page.
Your second image in addition has a so-called /SMask or /Mask, stroed under xref 86. So when you want to regenerate the original image, you must extract them both and use the mask image to set the transparency of the image under xref 87.
Have a look at [this](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/extract-imga.py) script for this.

**Comment by JorjMcKie** [comment]
To give you peace of mind, determine the images' rectangles on the page and, if (almost) equal to page.rect, simply make a page pixmap.
